src/Aspire.Hosting.Azure.AppService/AzureAppServiceEnvironmentExtensions.cs

```diff
@@ -0,0 +1,130 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting.Azure;
+using Aspire.Hosting.Azure.AppService;
+using Aspire.Hosting.Lifecycle;
+using Azure.Provisioning;
+using Azure.Provisioning.AppService;
+using Azure.Provisioning.ContainerRegistry;
+using Azure.Provisioning.Expressions;
+using Azure.Provisioning.Roles;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Aspire.Hosting;
+
+/// <summary>
+/// Extensions for adding Azure App Service Environment resources to a distributed application builder.
+/// </summary>
+public static partial class AzureAppServiceEnvironmentExtensions
+{
+    /// <summary>
+    /// Adds a azure app service environment resource to the distributed application builder.
+    /// </summary>
+    /// <param name="builder">The distributed application builder.</param>
+    /// <param name="name">The name of the resource.</param>
+    /// <returns><see cref="IResourceBuilder{T}"/></returns>
+    public static IResourceBuilder<AzureAppServiceEnvironmentResource> AddAppServiceEnvironment(this IDistributedApplicationBuilder builder, string name)
+    {
+        builder.AddAzureProvisioning();
+        builder.Services.Configure<AzureProvisioningOptions>(options => options.SupportsTargetedRoleAssignments = true);
+
+        if (builder.ExecutionContext.IsPublishMode)
+        {
+            builder.Services.TryAddLifecycleHook<AzureAppServiceInfrastructure>();
+        }
+
+        var resource = new AzureAppServiceEnvironmentResource(name, static infra =>
+        {
+            var prefix = infra.AspireResource.Name;
+            var resource = infra.AspireResource;
+
+            // This tells azd to avoid creating infrastructure
+            var userPrincipalId = new ProvisioningParameter(AzureBicepResource.KnownParameters.UserPrincipalId, typeof(string));
+            infra.Add(userPrincipalId);
+
+            var tags = new ProvisioningParameter("tags", typeof(object))
+            {
+                Value = new BicepDictionary<string>()
+            };
+
+            infra.Add(tags);
+
+            var identity = new UserAssignedIdentity(Infrastructure.NormalizeBicepIdentifier($"{prefix}-mi"))
+            {
+                Tags = tags
+            };
+
+            infra.Add(identity);
+
+            ContainerRegistryService? containerRegistry = null;
+            if (resource.TryGetLastAnnotation<ContainerRegistryReferenceAnnotation>(out var registryReferenceAnnotation) && registryReferenceAnnotation.Registry is AzureProvisioningResource registry)
+            {
+                containerRegistry = (ContainerRegistryService)registry.AddAsExistingResource(infra);
+            }
+            else
+            {
+                containerRegistry = new ContainerRegistryService(Infrastructure.NormalizeBicepIdentifier($"{prefix}_acr"))
+                {
+                    Sku = new() { Name = ContainerRegistrySkuName.Basic },
```

Comment:
I wonder if we should consolidate this across ACA and here, so there is only 1 place to change the defaults of an ACR.

Reply:
This, endpoint grouping, and env and arg processing.